File: 210322_BOJ_1213.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        char[] arr = br.readLine().toCharArray();

        int[] alpha = new int[26];

        for (char c : arr) {
            alpha[c - 'A']++;
        }

        int cnt = 0;
        StringBuilder mid = new StringBuilder();
        for(int i = 0; i < 26; ++i) {
            if (alpha[i] % 2 == 0)
                continue;

            mid.append((char)('A' + i));
            alpha[i]--;
            cnt++;
        }

        StringBuilder head = new StringBuilder();
        for(int i = 0; i < 26; ++i) {
            int Max = alpha[i] / 2;

            for(int j = 0; j < Max; ++j) {
                head.append((char)('A' + i));
                alpha[i]--;
            }
        }

        StringBuilder tail = new StringBuilder();
        for(int i = 25; i >= 0; --i) {
            while(alpha[i] > 0) {
                tail.append((char)('A' + i));
                alpha[i]--;
            }
        }

        if (cnt >= 2) {
            System.out.println("I'm Sorry Hansoo\n");
        }
        else {
            System.out.print(head);
            System.out.print(mid);
            System.out.println(tail);
        }
    }
}

File: 210322_BOJ_1213.cpp
#include <iostream>
#include <string>

using namespace std;

int alpha[26];

int main() {

	string t;
	cin >> t;

	char Max = 'A' - 1;
	for (int i = 0; i < t.size(); ++i) {

		alpha[t[i] - 'A']++;

		if (Max < t[i])
			Max = t[i];
	}

	string mid;
	for (int i = 0; i <= Max - 'A'; ++i) {
		if (alpha[i] == 0 || alpha[i] % 2 == 0) 
			continue;
		mid.push_back('A' + i);
		alpha[i]--;
	}
	
	if (mid.size() > 1) {
		cout << "I'm Sorry Hansoo\n";
		return 0;
	}

	string head;
	for (int i = 0; i <= Max - 'A'; ++i) {
		int cnt = alpha[i] / 2;

		for (int j = 0; j < cnt; ++j) {
			head.push_back('A' + i);
			alpha[i]--;
		}
	}

	string tail;
	for (int i = Max - 'A'; i >= 0; --i) {
		while (alpha[i]--) {
			tail.push_back('A' + i);
		}
	}

	cout << head << mid << tail << "\n";

	return 0;
}